README.md
understreck


HISTORY.rst
=======
History
=======

Unreleased
----------

1.0.0 (2020-03-03)
------------------
* Get elements from nested lists

0.5.1 (2019-11-09)
------------------
* Updates dev dependencies 

0.5.0 (2019-05-03)
------------------
* Updates dev dependencies
* Moves to the new repository

0.4.0 (2019-05-03)
------------------
* Adds Understreck.strip that strips indents

0.3.0 (2019-05-03)
------------------
* Adds Understreck.filter with inspired by https://lodash.com/docs/4.17.11#filter
* Updated the travis button URL
* Replaced nested_get with get in the README. (nested_get still works)
* Black formatting

0.2.1 (2018-12-10)
------------------
* Add Understreck.chunks

0.2.0 (2018-11-12)
------------------
* Add get as an alias for nested get

0.1.1 (2018-11-02)
------------------
* Update the readme

0.1.0 (2018-11-02)
------------------

* First release on PyPI.


tests/test_chunks.py
from __future__ import absolute_import

import understreck as _


def test_chunk_size():
    to_chunk = ["one", "two", "three", "four", "five"]
    chunk_size = _.chunks.calculate_size(to_chunk, 2)
    assert chunk_size == 3
    assert isinstance(chunk_size, int)


def test_chunks_5():
    to_chunk = ["one", "two", "three", "four", "five"]
    result = _.chunks.split(to_chunk, 2)
    expected = [["one", "two", "three"], ["four", "five"]]
    assert result == expected


def test_chunks_7():
    to_chunk = ["one", "two", "three", "four", "five", "six", "seven"]
    result = _.chunks.split(to_chunk, 2)
    expected = [["one", "two", "three", "four"], ["five", "six", "seven"]]
    assert result == expected


tests/test_filter.py
from __future__ import absolute_import

import understreck as _


def test_filter__function():
    users = [
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    ]

    result = _.filter(users, lambda x: not x.get("active"))
    assert result == [{"user": "fred", "age": 40, "active": False}]


def test_filter__function_tuple():
    users = (
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    )

    result = _.filter(users, lambda x: not x.get("active"))
    assert result == ({"user": "fred", "age": 40, "active": False},)


def test_filter__matches():
    users = [
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    ]

    result = _.filter(users, {"age": 36, "active": True})
    assert result == [{"user": "barney", "age": 36, "active": True}]


def test_filter__matches_tuple():
    users = (
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    )

    result = _.filter(users, {"age": 36, "active": True})
    assert result == ({"user": "barney", "age": 36, "active": True},)


def test_filter__matches_property():
    users = [
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    ]

    result = _.filter(users, ["active", False])
    assert result == [{"user": "fred", "age": 40, "active": False}]


def test_filter__property():
    users = [
        {"user": "barney", "age": 36, "active": True},
        {"user": "fred", "age": 40, "active": False},
    ]

    result = _.filter(users, ["active"])
    assert result == [{"user": "barney", "age": 36, "active": True}]


tests/test_understreck.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `understreck` package."""
from __future__ import absolute_import
import pytest


import understreck as _

test_dictionary = {
    "the_top_level": {
        "second_level": {"third_level": "it works", "third_level_sibling": False}
    }
}


@pytest.mark.parametrize(
    "dict_to_test,key_to_get,expected",
    [
        (test_dictionary, "the_top_level.second_level.third_level", "it works"),
        (test_dictionary, "the_top_level.second_level.third_level_sibling", False),
        (test_dictionary, "the_top_level.second_level.DOES_NOT_EXIST", None),
        (test_dictionary, ["the_top_level", "second_level", "DOES_NOT_EXIST"], None),
        (
            test_dictionary,
            "the_top_level.second_level",
            {"third_level": "it works", "third_level_sibling": False},
        ),
        (test_dictionary, ["the_top_level", "second_level", "third_level"], "it works"),
        (
            test_dictionary,
            ["the_top_level", "second_level"],
            {"third_level": "it works", "third_level_sibling": False},
        ),
    ],
)
def test_nested_get(dict_to_test, key_to_get, expected):
    result = _.nested_get(dict_to_test, key_to_get)
    assert result == expected
    result = _.get(dict_to_test, key_to_get)
    assert result == expected


@pytest.mark.parametrize(
    "dict_to_test,key_to_get,expectedException",
    [
        (None, "", _.exceptions.InvalidArgumentError),
        ("", "", _.exceptions.InvalidArgumentError),
        (1, "", _.exceptions.InvalidArgumentError),
        (test_dictionary, None, _.exceptions.InvalidArgumentError),
        (test_dictionary, {}, _.exceptions.InvalidArgumentError),
        (test_dictionary, 1, _.exceptions.InvalidArgumentError),
    ],
)
def test_nested_get_exceptions(dict_to_test, key_to_get, expectedException):
    print(expectedException)
    with pytest.raises(expectedException):
        _.nested_get(dict_to_test, key_to_get)


def test_get_elements_from_list():
    test_dictionary = {"the_top_level": {"second_level_list": ["Hello", "World"]}}
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[0]")
    assert result == "Hello"
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[0]")
    assert result == "Hello"

    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[1]")
    assert result == "World"
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[1]")
    assert result == "World"


def test_get_elements_from_list__object_in_list():
    test_dictionary = {
        "the_top_level": {"second_level_list": [{"name": "Kalle"}, "World"]}
    }
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[0]")
    assert result == {"name": "Kalle"}
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[0]")
    assert result == {"name": "Kalle"}

    assert result == {"name": "Kalle"}
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[0].name")
    assert result == "Kalle"

    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[1]")
    assert result == "World"
    result = _.nested_get(test_dictionary, "the_top_level.second_level_list[1]")
    assert result == "World"


def test_get_elements_from_list__2_levels_deep():
    test_dictionary = {
        "the_top_level": {
            "second_level_list": [
                {"tags": [{"name": "first_tag"}, {"name": "second_tag"}]}
            ]
        }
    }

    result = _.get(test_dictionary, "the_top_level.second_level_list[0].tags[1].name")
    assert result == "second_tag"


def test_get_elements_from_list__3_levels_deep():
    test_dictionary = {
        "the_top_level": {
            "second_level_list": [
                {
                    "tags": [
                        {"name": "first_tag"},
                        {"name": "second_tag"},
                        [{"name": "first_tag"}, {"name": "second_tag"}],
                    ]
                }
            ]
        }
    }

    result = _.get(test_dictionary, "the_top_level.second_level_list[0].tags[2]")
    assert result == [{"name": "first_tag"}, {"name": "second_tag"}]

    result = _.get(
        test_dictionary, "the_top_level.second_level_list[0].tags[2][1].name"
    )
    assert result == "second_tag"


def test_get_elements_from_list__deep_1():
    test_dictionary = {"foo": {"bar": ["x", ["first", "second", "Hello World"]]}}

    result = _.get(test_dictionary, "foo.bar[1][2]",)
    assert result == "Hello World"


def test_get_elements_from_list__deep_2():
    test_dictionary = {
        "foo": {"bar": ["x", ["first", "second", {"name": "Hello World"}]]}
    }

    result = _.get(test_dictionary, "foo.bar[1][2].name",)
    assert result == "Hello World"


def test_get_elements_from_list__deep_3():
    test_dictionary = {"foo": {"bar": ["x", "y", ["x", "y", "z", "Hello World",]]}}

    result = _.get(test_dictionary, "foo.bar[2][3]",)
    assert result == "Hello World"


def test_get_elements_from_list__deep_4():
    test_dictionary = {
        "data": {
            "signUp": {
                "somethingElese": "hello",
                "errors": [{"name": "Error name", "field": "Hello World"}],
            }
        }
    }

    result = _.get(test_dictionary, "data.signUp.errors[0].field",)
    assert result == "Hello World"


understreck/chunks.py
import math


def calculate_size(to_chunk, num_chunks):
    """Calculate the max chunk size for each new chunk of a list.

    Arguments:
        to_chunk {list} -- [description]
        num_chunks {int} -- [description]

    Returns:
        int -- The max size for each chunk
    """

    chunk_max_size = float(len(to_chunk)) / float(num_chunks)
    return int(math.ceil(chunk_max_size))


def split(to_chunk, num_chunks):
    """ split a list into evenly sized chunks

    Args:
        to_chunk: The original list to turn in to chunks
        num_chunks: The total number of chunks to create

    Returns: A new list with the

    """
    chunk_max_size = calculate_size(to_chunk, num_chunks)
    chunked_list = [
        to_chunk[i : i + chunk_max_size]
        for i in range(0, len(to_chunk), chunk_max_size)
    ]

    return chunked_list


understreck/nested_get.py
# -*- coding: utf-8 -*-
from .exceptions import InvalidArgumentError


def _is_list(key):
    return "[" in key


def _strip_position(key):
    return key.split("[")[0]


def _get_positions(key):
    key_without_name = key.split("[")[0]
    positions_string = key.replace(key_without_name, "")
    raw_positions = positions_string.split("]")
    raw_positions = [x for x in raw_positions if x != ""]

    positions = [_position(x) for x in raw_positions]
    return positions


def _position(key):
    position = key.split("[")[-1]
    position = position.split("]")[0]
    return int(position)


def nested_get(input_dict, nested_key):
    """Get a subkey from a dict

    This will return a subkey from a dictionary.
    It solves a problem when having to perform safe get on nested dictionaries.

    Arguments:
        input_dict {dict} -- The dictionary to perform the get on
        nested_key {str, list, tuple} -- The key can be or dot delimited string
                                         or list/tuple.

    Returns:
        Object or None -- The value of the key or None
    """
    if not isinstance(nested_key, (str, list, tuple)):
        raise InvalidArgumentError("The nested_key is not the correct type")
    if not isinstance(input_dict, dict):
        raise InvalidArgumentError("The input_dict is not the correct type")

    normalized_nested_key = None
    if isinstance(nested_key, str):
        normalized_nested_key = nested_key.split(".")
    elif isinstance(nested_key, (list, tuple)):
        normalized_nested_key = nested_key

    internal_dict_value = input_dict
    for k in normalized_nested_key:
        if _is_list(k):
            # Handle chained lists
            positions = _get_positions(k)
            internal_dict_value = internal_dict_value.get(_strip_position(k), None)
            for position in positions:
                try:
                    internal_dict_value = internal_dict_value[position]
                except IndexError:
                    return None
        else:
            internal_dict_value = internal_dict_value.get(k, None)
        if internal_dict_value is None:
            return None
    return internal_dict_value


get = nested_get


understreck/understreck.py
# -*- coding: utf-8 -*-


def nested_get(input_dict, nested_key):
    """Get a subkey from a dict

    This will return a subkey from a dictionary. 
    It solves a problem when having to perform safe get on nested dictionaries.
    
    Arguments:
        input_dict {dict} -- The dictionary to perform the get on
        nested_key {str, list, tuple} -- The key can be or dot delimited string or list/tuple.
    
    Returns:
        Object or None -- The value of the key or None
    """
    if not isinstance(nested_key, (str, list, tuple)):
        raise InvalidArgumentError("The nested_key is not the correct type")
    if not isinstance(input_dict, dict):
        raise InvalidArgumentError("The input_dict is not the correct type")

    normalized_nested_key = None
    if isinstance(nested_key, str):
        normalized_nested_key = nested_key.split(".")
    elif isinstance(nested_key, (list, tuple)):
        normalized_nested_key = nested_key

    internal_dict_value = input_dict
    for k in normalized_nested_key:
        internal_dict_value = internal_dict_value.get(k, None)
        if internal_dict_value is None:
            return None
    return internal_dict_value


class InvalidArgumentError(Exception):
    pass


understreck/filter.py
def filter(to_filter, filter_to_apply):
    """Filters a list or tuple

    Examples:
        users = [
                {"user": "barney", "age": 36, "active": True},
                {"user": "fred", "age": 40, "active": False},
            ]

        # Using a lambda function
        result = _.filter(users, lambda x: not x.get("active"))
        result == [{"user": "fred", "age": 40, "active": False}]

        # Using partial dictionary
        result = _.filter(users, {"age": 36, "active": True})
        result == [{"user": "barney", "age": 36, "active": True}]

        # Using a list with a property name and value
        result = _.filter(users, ["active", False])
        result == [{"user": "fred", "age": 40, "active": False}]

        # Using a list with a property name. The value must be truthy.
        result = _.filter(users, ["active"])
        result == [{"user": "barney", "age": 36, "active": True}]

    :param to_filter: the tuple or list to filter
    :param filter_to_apply: A function or list that defies the filter
    :return: A filtered list or tuple with the same type as to_filter
    """
    type_to_return = type(to_filter)

    if hasattr(filter_to_apply, "__call__"):
        return type_to_return([x for x in to_filter if filter_to_apply(x)])

    if isinstance(filter_to_apply, dict):
        filtered = []
        for x in to_filter:
            matched = 0
            keys_in_filter = filter_to_apply.keys()
            total_to_find = len(keys_in_filter)
            for key in keys_in_filter:
                if filter_to_apply[key] == x[key]:
                    matched += 1

            if total_to_find == matched:
                filtered.append(x)

        return type_to_return(filtered)

    if isinstance(filter_to_apply, list):
        if len(filter_to_apply) == 2:
            property_name = filter_to_apply[0]
            property_value = filter_to_apply[1]
            result = [x for x in to_filter if x.get(property_name) == property_value]
            return type_to_return(result)

        if len(filter_to_apply) == 1:
            property_name = filter_to_apply[0]
            result = [x for x in to_filter if bool(x.get(property_name))]
            return type_to_return(result)
